File: platforms/android/samples/CheckoutKitAndroidDemo/app/src/main/java/com/shopify/checkout_kit_android_demo/e2e/CartBootstrap.kt
package com.shopify.checkout_kit_android_demo.e2e

import android.net.Uri

internal data class CartBootstrapRequest(
    val productIndex: Int,
    val quantity: Int,
)

internal object CartBootstrap {
    private const val SCHEME = "com.shopify.checkoutkit.androiddemo"
    private const val HOST = "cart"

    fun request(uri: Uri): CartBootstrapRequest? {
        if (uri.scheme?.lowercase() != SCHEME) return null

        require(uri.host?.lowercase() == HOST && (uri.path.isNullOrEmpty() || uri.path == "/")) {
            "Unsupported cart bootstrap route"
        }
        require(uri.isHierarchical) { "Invalid cart bootstrap URL" }

        return CartBootstrapRequest(
            productIndex = positiveInteger(uri, "productIndex", allowingZero = true),
            quantity = positiveInteger(uri, "quantity", allowingZero = false),
        )
    }

    private fun positiveInteger(uri: Uri, name: String, allowingZero: Boolean): Int {
        val value = uri.getQueryParameters(name).singleOrNull()?.toIntOrNull()
        require(value != null && if (allowingZero) value >= 0 else value > 0) {
            "Invalid cart bootstrap parameter: $name"
        }
        return value
    }
}

File: platforms/swift/Samples/CheckoutKitSwiftDemo/CheckoutKitSwiftDemo/Sources/E2E/CartBootstrap.swift
import Foundation

struct CartBootstrapRequest: Sendable {
    let productIndex: Int
    let quantity: Int
}

enum CartBootstrap {
    private static let scheme = "com.shopify.checkoutkit.swiftdemo"
    private static let host = "cart"

    static func request(from url: URL) throws -> CartBootstrapRequest? {
        guard url.scheme?.lowercased() == scheme else {
            return nil
        }

        guard url.host?.lowercased() == host, url.path.isEmpty || url.path == "/" else {
            throw ValidationError.unsupportedRoute
        }

        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw ValidationError.invalidURL
        }

        return try CartBootstrapRequest(
            productIndex: positiveInteger(named: "productIndex", in: components, allowingZero: true),
            quantity: positiveInteger(named: "quantity", in: components, allowingZero: false)
        )
    }

    private static func positiveInteger(
        named name: String,
        in components: URLComponents,
        allowingZero: Bool
    ) throws -> Int {
        let values = components.queryItems?.filter { $0.name == name } ?? []
        guard values.count == 1, let rawValue = values.first?.value, let value = Int(rawValue) else {
            throw ValidationError.invalidParameter(name: name)
        }

        guard allowingZero ? value >= 0 : value > 0 else {
            throw ValidationError.invalidParameter(name: name)
        }

        return value
    }

    enum ValidationError: LocalizedError, Sendable {
        case invalidURL
        case unsupportedRoute
        case invalidParameter(name: String)

        var errorDescription: String? {
            switch self {
            case .invalidURL:
                return "The cart bootstrap URL is invalid."
            case .unsupportedRoute:
                return "The cart bootstrap route is unsupported."
            case let .invalidParameter(name):
                return "The cart bootstrap parameter \(name) is invalid."
            }
        }
    }
}
